src/main/java/rx/internal/operators/OnSubscribeRedo.java

```diff
@@ -235,8 +235,9 @@ public void onError(Throwable e) {
                     @Override
                     public void onNext(T v) {
                         if (!done) {
-                            if (consumerCapacity.get() != Long.MAX_VALUE) {
-                                consumerCapacity.decrementAndGet();
+                            long cap = consumerCapacity.get();
```

Comment:
Generally, one should decrement the request/capacity after the value has been emitted, because otherwise if the value becomes zero, it may open a window for a concurrent emission that is kicked off by a transition from 0 to some value. Could you see if that helps  with #2863 ?